File: src/parallel_you/storage/backends/mongo.py
import re
from datetime import datetime
from pymongo import MongoClient, ReturnDocument

from parallel_you.model import Story, Saga, Thread
from parallel_you.storage import Repo
from filter_spec import FilterSpec

class MongoRepo(Repo):
    def __init__(self, uri: str = "mongodb://127.0.0.1:27017/parallel_you"):
        if not uri:
            raise ValueError("Mongo URI must be provided")
        client = MongoClient(uri)
        db = client.get_default_database()
        if db is None:
            raise ValueError("Mongo URI must include a database name, e.g. .../parallel_you")
        self._c = db.get_collection("threads")
        # Helpful indexes
        self._c.create_index("title")
        self._c.create_index("archived")
        self._c.create_index("kind")

    # ---------- mapping ----------
    def _to_doc(self, t: Thread) -> dict:
        # Persist a discriminator for queries
        if isinstance(t, Saga):
            kind = "saga"
            extra = {"stories": []}  # or omit if not embedding
        elif isinstance(t, Story):
            kind = "story"
            extra = {
                "saga_id": t.saga_id,
                "state": t.state.name,
                "energy_band": t.energy.name,
                "quantum_overridden": getattr(t, "quantum_overridden", False),
            }
        else:
            kind = "thread"
            extra = {}

        return {
            "_id": t.id,                      # string id
            "kind": kind,
            "title": t.title,
            "priority": t.priority,
            "quantum": t.quantum,
            "archived": t.archived,
            "created_at": t.created_at,
            "updated_at": t.updated_at,
            "archived_at": t.archived_at,
            **extra,
        }

    def _from_doc(self, d: dict) -> Thread:
        kind = d.get("kind", "thread")
        base = dict(
            id=d.get("_id"),
            title=d["title"],
            priority=int(d.get("priority", 1)),
            quantum=d.get("quantum", "50m"),
            archived=bool(d.get("archived", False)),
            created_at=d.get("created_at"),
            updated_at=d.get("updated_at"),
            archived_at=d.get("archived_at"),
        )
        if kind == "saga":
            return Saga(**base, stories=[])
        if kind == "story":
            # adapt state/energy parsing if you store names vs values
            from parallel_you.model.story_state import StoryState
            from parallel_you.model.energy_band import EnergyBand
            state = d.get("state")
            energy = d.get("energy_band")
            return Story(
                saga_id=d["saga_id"],
                state=StoryState[state] if isinstance(state, str) else state,
                energy=EnergyBand[energy] if isinstance(energy, str) else energy,
                **base,
            )
        return Thread(**base)

    # ---------- repo API ----------
    def list(self, spec: FilterSpec) -> list[Thread]:
        q: dict = {}
        if spec.archived is not None:
            q["archived"] = spec.archived
        if spec.text:
            q["title"] = {"$regex": re.escape(spec.text), "$options": "i"}
        if spec.types:
            q["kind"] = {"$in": [cls.__name__.lower() for cls in spec.types]}
        return [self._from_doc(d) for d in self._c.find(q)]

    def get(self, id: str) -> Thread | None:
        d = self._c.find_one({"_id": id})
        return self._from_doc(d) if d else None

    def upsert(self, t: Thread) -> str:
        # Read existing to preserve created_at and handle archived_at transitions
        existing = self._c.find_one({"_id": t.id})
        now = datetime.now().replace(second=0, microsecond=0)

        created_at = (existing or {}).get("created_at") or getattr(t, "created_at", None) or now

        prev_arch = bool((existing or {}).get("archived", False))
        new_arch = bool(getattr(t, "archived", prev_arch))
        if not prev_arch and new_arch:
            archived_at = now
        elif prev_arch and not new_arch:
            archived_at = None
        else:
            archived_at = (existing or {}).get("archived_at", getattr(t, "archived_at", None))

        # Build doc snapshot
        doc = self._to_doc(t)
        doc["created_at"] = created_at
        doc["updated_at"] = now
        doc["archived_at"] = archived_at

        self._c.replace_one({"_id": t.id}, doc, upsert=True)
        return t.id

    def delete(self, id: str) -> bool:
        return self._c.delete_one({"_id": id}).deleted_count == 1

    def clear(self) -> None:
        self._c.delete_many({})

File: src/ParallelYou/ParallelYou.Repo/Backends/Mongo.cs
using MongoDB.Bson;
using MongoDB.Driver;
using ParallelYou.Repo.Abstractions;
using VyThread = ParallelYou.Model.Threads.Thread;
using ParallelYou.Model.Threads;

namespace ParallelYou.Repo.Backends;

// Minimal MongoDB implementation of IRepo.
// Stores all Thread-derived types in a single collection with a discriminator field "_t".
public class MongoRepo : IRepo
{
    private readonly IMongoCollection<BsonDocument> _col;

    public MongoRepo(string mongoUri, string databaseName = "parallel_you", string collectionName = "threads")
    {
        var db = new MongoClient(mongoUri).GetDatabase(databaseName);
        _col = db.GetCollection<BsonDocument>(collectionName);
    }

    public IList<VyThread> List(FilterSpec filter)
    {
        var fb = Builders<BsonDocument>.Filter;
        var filters = new List<FilterDefinition<BsonDocument>>();

        if (!string.IsNullOrWhiteSpace(filter.Text))
        {
            var text = filter.Text.Trim();
            // Case-insensitive contains on title
            filters.Add(fb.Regex("title", new BsonRegularExpression(text, "i")));
        }

        if (filter.Archived is not null)
        {
            filters.Add(fb.Eq("archived", filter.Archived.Value));
        }

        if (filter.Types is { Count: > 0 })
        {
            var names = filter.Types.Select(TypeToDiscriminator).ToArray();
            filters.Add(fb.In("_t", names));
        }

        var mongoFilter = filters.Count switch
        {
            0 => FilterDefinition<BsonDocument>.Empty,
            1 => filters[0],
            _ => fb.And(filters)
        };

        var docs = _col.Find(mongoFilter).ToList();
        return docs.Select(Deserialize).Select(t => t.Clone()).ToList();
    }

    public VyThread Get(string itemId)
    {
        var doc = _col.Find(Builders<BsonDocument>.Filter.Eq("_id", itemId)).FirstOrDefault();
        if (doc is null) throw new KeyNotFoundException($"No thread with id '{itemId}'");
        return Deserialize(doc).Clone();
    }

    public string Upsert(VyThread item)
    {
        var doc = Serialize(item);
        var filter = Builders<BsonDocument>.Filter.Eq("_id", item.Id);
        _col.ReplaceOne(filter, doc, new ReplaceOptions { IsUpsert = true });
        return item.Id;
    }

    public bool Delete(string itemId)
    {
        var res = _col.DeleteOne(Builders<BsonDocument>.Filter.Eq("_id", itemId));
        return res.DeletedCount > 0;
    }

    public void Clear()
    {
        _col.DeleteMany(FilterDefinition<BsonDocument>.Empty);
    }

    private static string TypeToDiscriminator(Type t)
    {
        if (t == typeof(Domain)) return nameof(Domain);
        if (t == typeof(Saga)) return nameof(Saga);
        if (t == typeof(Story)) return nameof(Story);
        // Fallback to simple name
        return t.Name;
    }

    private static BsonDocument Serialize(VyThread t)
    {
        var doc = new BsonDocument
        {
            ["_id"] = t.Id,
            ["title"] = t.Title,
            ["priority"] = t.Priority,
            ["quantum"] = t.Quantum,
            ["archived"] = t.Archived,
            ["createdAt"] = t.CreatedAt.HasValue ? (BsonValue) t.CreatedAt.Value : BsonNull.Value,
            ["updatedAt"] = t.UpdatedAt.HasValue ? (BsonValue) t.UpdatedAt.Value : BsonNull.Value,
            ["archivedAt"] = t.ArchivedAt.HasValue ? (BsonValue) t.ArchivedAt.Value : BsonNull.Value,
        };

        switch (t)
        {
            case Domain d:
                doc["_t"] = nameof(Domain);
                if (d.Description != null) doc["description"] = d.Description;
                break;
            case Saga s:
                doc["_t"] = nameof(Saga);
                doc["domainId"] = s.DomainId;
                break;
            case Story st:
                doc["_t"] = nameof(Story);
                doc["sagaId"] = st.SagaId;
                doc["dueAt"] = st.DueAt.HasValue ? (BsonValue) st.DueAt.Value : BsonNull.Value;
                doc["energy"] = (int)st.Energy;
                doc["storyState"] = (int)st.StoryState;
                break;
            default:
                doc["_t"] = t.GetType().Name;
                break;
        }

        return doc;
    }

    private static VyThread Deserialize(BsonDocument doc)
    {
        var type = doc.GetValue("_t", BsonValue.Create("Thread")).AsString;

        string id = doc["_id"].AsString;
        string title = doc.GetValue("title", "").AsString;
        int priority = doc.GetValue("priority", 0).ToInt32();
        string quantum = doc.GetValue("quantum", "").AsString;
        bool archived = doc.GetValue("archived", false).ToBoolean();
        DateTime? createdAt = doc.Contains("createdAt") && doc["createdAt"].IsBsonNull == false ? doc["createdAt"].ToUniversalTime() : null;
        DateTime? updatedAt = doc.Contains("updatedAt") && doc["updatedAt"].IsBsonNull == false ? doc["updatedAt"].ToUniversalTime() : null;
        DateTime? archivedAt = doc.Contains("archivedAt") && doc["archivedAt"].IsBsonNull == false ? doc["archivedAt"].ToUniversalTime() : null;

        return type switch
        {
            nameof(Domain) => new Domain(
                id: id,
                title: title,
                priority: priority,
                quantum: quantum,
                description: doc.GetValue("description", BsonNull.Value).IsBsonNull ? null : doc["description"].AsString,
                archived: archived,
                createdAt: createdAt,
                updatedAt: updatedAt,
                archivedAt: archivedAt
            ),
            nameof(Saga) => new Saga(
                id: id,
                domainId: doc.GetValue("domainId", "").AsString,
                title: title,
                priority: priority,
                quantum: quantum,
                archived: archived,
                createdAt: createdAt,
                updatedAt: updatedAt,
                archivedAt: archivedAt
            ),
            nameof(Story) => new Story(
                id: id,
                sagaId: doc.GetValue("sagaId", "").AsString,
                title: title,
                priority: priority,
                quantum: quantum,
                energy: (EnergyBand) doc.GetValue("energy", 0).ToInt32(),
                state: (StoryState) doc.GetValue("storyState", 0).ToInt32(),
                archived: archived,
                dueAt: doc.Contains("dueAt") && doc["dueAt"].IsBsonNull == false ? doc["dueAt"].ToUniversalTime() : null,
                createdAt: createdAt,
                updatedAt: updatedAt,
                archivedAt: archivedAt
            ),
            _ => throw new InvalidOperationException($"Unknown thread type '{type}'")
        };
    }
}
